### PR TITLE
Support running workouts in schedule_workout / save_workout_template

### PR DESCRIPTION
**Pull request by LeonLukaschek**
## Summary

`schedule_workout` / `save_workout_template` previously only produced valid payloads for cycling/intervals. Passing a run (e.g. `sport_type=100`) produced an entry that crashed the COROS iPhone app on open, and once enough fields were stripped to stop the crash the watch rendered the entry as Strength (no running interface, no HR zones).

This PR makes the same builder produce a valid Running payload.

## What changed

`_build_workout_program_payload` in `coros_api.py`:

- **Sport-type namespace mapping.** The COROS activity API and workout API use different sport IDs (activity `100` = Running, workout `1` = Running). Callers naturally pass the activity-side ID (it's what `list_activities` returns). The builder now accepts `sport_type=100` and maps it to wire `sportType=1` at program and exercise level. Cycling IDs are passed through unchanged.
- **Structured-workout metadata block.** Run programs need the same metadata strength programs already carry — without it the COROS app fails to parse the entry. Added at program level: `subType=65535` (the structured-workout marker), `referExercise`, `type`, `overview`, `poolLength*`, `sourceUrl`, `videoCoverUrl/Url`, `gradeSystemVersion`, `hybridTotalSets`, `duration`, `exerciseNum`, `totalSets`, `trainingLoad`. Added per step: `exerciseKind`, `gradeSystem`, `hrType`, `intensityMultiplier`, `intensityPercent*`, `onsightGradeOffset`, `overview`, `packageTime`, `sourceId`, `subType`, `targetDisplayUnit`.
- **Per-step `exerciseType` by position:** `1`=warmup (first step), `3`=cooldown (last step), `2`=main (everything else, and single-step workouts).
- **HR targeting:** when `intensity_type=2` (HR), per-step `hrType=2` and program-level `referExercise.hrType=3`.
- **`WORKOUT_SPORT_NAMES` fix:** the previous entry mapped Running to `100`, but the workout API actually returns `sportType=1`, so it never round-tripped. Replaced with `1: \"Running\"`.

Cycling/intervals path is untouched.

## Tests

`tests/test_workout_payloads.py` — 6 new tests:

- Namespace mapping (`sport_type=100` → wire `1`).
- Program-level structured-workout metadata block present for runs.
- Per-step `exerciseType` varies by position (`[1, 2, 3]`).
- Single-step run uses `exerciseType=2`.
- HR intensity sets per-step `hrType=2` + `referExercise.hrType=3`.
- Cycling payloads unchanged.

Full suite (81 tests) passes.

## Manual verification

Pushed an HR-targeted multi-step run via the patched builder:

- Opens in the COROS iPhone app
- Syncs to the watch
- Renders as Running with the prescribed HR zones per step

### CODE REVIEW

File: coros_mcp/coros_api.py [hunk 1]
```diff
@@ -725,14 +735,68 @@ def _build_workout_program_payload(
...
+    if is_running:
+        # Per-step exerciseType: 1=warmup, 2=main, 3=cooldown.
+        # hrType=2 marks HR-based targets.
+        plain_exercises = [e for e in exercises if not e.get("isGroup")]
+        last_idx = len(plain_exercises) - 1
+        for i, ex in enumerate(plain_exercises):
+            if i == 0 and last_idx > 0:
+                ex["exerciseType"] = 1   # warmup
+            elif i == last_idx and last_idx > 0:
+                ex["exerciseType"] = 3   # cooldown
```
Comment (cygnusb) [inline coros_mcp/coros_api.py]: **Bug: misclassifies steps inside repeat groups.**

`plain_exercises` filters out only `isGroup=True` rows, but sub-steps of a repeat group have `isGroup=False` (with `groupId != "0"`). For a common run-interval shape like:

```
[ repeat(3× [hard, easy]) ]
```

`plain_exercises` becomes `[hard, easy]`, so `hard` is tagged `exerciseType=1` (warmup) and `easy` is tagged `3` (cooldown) — both wrong; both should be `2` (main).

Even for `[plain warmup, repeat(4× [hard, easy]), plain cooldown]` you get `[warmup, hard, easy, cooldown]`: the warmup/cooldown positions are right by luck, but if the user puts extra plain steps around a group the labels drift.

Suggested fix: restrict warmup/cooldown markers to *top-level* plain steps only — iterate over the original `steps` list and assign markers to the first/last entries that are plain (`"repeat" not in step`). Sub-steps inside a group should always be `exerciseType=2`.

File: coros_api.py [hunk 2]
```diff
@@ -725,14 +735,68 @@ def _build_workout_program_payload(
...
         "exercises": exercises,
     }
 
+    # Running programs need the same metadata block strength programs
+    # carry. Without it the COROS app fails to parse the entry or renders
+    # it as strength on the watch.
+    if is_running:
+        # Per-step exerciseType: 1=warmup, 2=main, 3=cooldown.
+        # hrType=2 marks HR-based targets.
+        plain_exercises = [e for e in exercises if not e.get("isGroup")]
```
Comment (cygnusb) [inline coros_api.py]: Related to the bug on line 758: the predicate should be `not e.get("isGroup") and e.get("groupId", "0") == "0"` if you want "top-level plain steps only." As written, this also pulls in repeat sub-steps.

File: coros_api.py [hunk 3]
```diff
@@ -647,6 +651,12 @@ def _build_workout_program_payload(
     total_seconds = 0
     ex_id = 0  # sequential exercise IDs (API uses these to link groups)
 
+    # Workout API uses sportType=1 for Running; activity API uses
+    # sportType=100. Accept the activity-side ID (matches list_activities)
+    # and map to the workout-side ID for the payload.
+    is_running = sport_type == 100
```
Comment (cygnusb) [inline coros_api.py]: **Namespace coverage gap.** `SPORT_NAMES` (lines 475–480) shows multiple run-flavored activity IDs:

- `100` Running
- `102` Trail Running
- `103` Track Running

Only `100` triggers `is_running` here. `102` / `103` will silently produce a running-shaped payload **without** the metadata block — reproducing the exact failure mode this PR is fixing (app crash / strength rendering). The activity namespace also likely has an indoor-running / treadmill ID that should be covered.

At minimum, either:
- Expand `is_running` to `sport_type in {100, 102, 103, …}` (verify which workout-side `sportType` each maps to — they may all map to wire `1`, or treadmill may have its own ID), **or**
- Raise `ValueError` for unsupported run-flavored IDs with a clear message pointing at `100`.

Silent fall-through is the worst option, because the symptom (app crash) is far from the cause.

File: coros_mcp/coros_api.py [hunk 4]
```diff
@@ -647,6 +651,12 @@ def _build_workout_program_payload(
     total_seconds = 0
     ex_id = 0  # sequential exercise IDs (API uses these to link groups)
 
+    # Workout API uses sportType=1 for Running; activity API uses
+    # sportType=100. Accept the activity-side ID (matches list_activities)
+    # and map to the workout-side ID for the payload.
+    is_running = sport_type == 100
+    wire_sport_type = 1 if is_running else sport_type
```
Comment (cygnusb) [inline coros_mcp/coros_api.py:803]: **Leaky abstraction.** `is_running` is gated on `sport_type == 100` only, so passing `sport_type=1` (the wire ID a curious caller might try after reading `WORKOUT_SPORT_NAMES`) produces a running payload *without* the metadata block — same crash mode. Pick one:

- Accept both: `is_running = sport_type in (1, 100)`, then `wire_sport_type = 1 if is_running else sport_type`.
- Reject `1`: raise `ValueError("Pass sport_type=100 for running (activity-namespace ID); 1 is the internal workout-API ID and should not be used directly.")`.

Either is fine; silent divergence is not.

File: coros_mcp/coros_api.py [hunk 5]
```diff
@@ -583,9 +583,9 @@ async def fetch_activity_detail(auth: StoredAuth, activity_id: str, sport_type:
 # IntensityType values: 1=weight, 2=HR, 3=pace, 4=speed, 5=none, 6=power, 7=cadence
 
 WORKOUT_SPORT_NAMES: dict[int, str] = {
+    1: "Running",
```
Comment (cygnusb) [inline coros_mcp/coros_api.py:632]: Good catch on the workout-namespace mismatch — the old `100: "Running"` never round-tripped. Worth a one-line comment here to spare the next reader the same investigation:

```python
# Note: the workout API uses sportType=1 for Running; the activity API
# uses 100. _build_workout_program_payload maps 100 → 1 on the way out.
```

File: coros_mcp/coros_api.py [hunk 6]
```diff
@@ -635,10 +635,14 @@ def _build_workout_program_payload(
...
-    """Sync builder for the cycling/intervals program dict.
+    """Sync builder for the cycling/intervals/running program dict.
 
     steps: list of dicts — either plain steps or repeat groups (see
     save_workout_template docstring).
+
+    sport_type=100 builds a running program (mapped to the workout-side
+    sport ID and given the metadata block COROS requires for runs).
+    Cycling (the default) is unchanged.
     """
```
Comment (cygnusb) [inline coros_mcp/coros_api.py:778]: Worth expanding this docstring to spell out the supported `sport_type` values (and the namespace mismatch) — this is the canonical place a maintainer will look:

- 2 = Indoor Cycling (default)
- 100 = Running (activity-namespace ID; mapped to wire `sportType=1`)
- 200 = Road Bike, 201 = Indoor Cycling (alt)

And if you extend `is_running` to cover Trail (102) / Track (103) / Treadmill (see comment on line 657), document those here too.

File: tests/test_workout_payloads.py [hunk 7]
```diff
@@ -331,6 +331,106 @@ def test_cycling_sport_and_intensity_types_propagate():
...
+def test_running_does_not_affect_cycling():
+    """Cycling programs keep their existing sparse shape (no metadata block)."""
+    payload = _build_workout_program_payload(
+        name="c",
+        steps=[{"name": "ride", "duration_minutes": 30, "intensity_low": 200, "intensity_high": 250}],
+        sport_type=2,
+    )
+    assert payload["sportType"] == 2
+    assert "subType" not in payload
+    assert "referExercise" not in payload
```
Comment (cygnusb) [inline tests/test_workout_payloads.py:430]: **Test coverage gap:** no test exercises a running workout containing a repeat group. That's the common shape for intervals and where the warmup/cooldown classification bug surfaces (see comment on `coros_api.py:758`).

Suggested additions:

1. Run with `[plain warmup, repeat(3× [hard, easy]), plain cooldown]` → assert only the warmup step has `exerciseType=1`, only the cooldown has `3`, and **all sub-steps inside the group** have `2`.
2. Run with `[repeat(3× [hard, easy])]` only → assert both sub-steps have `exerciseType=2` (not 1 / 3).
3. Positive regression for cycling: assert the cycling payload's top-level key set equals exactly `{"name", "sportType", "estimatedTime", "access", "exercises"}` to catch accidental metadata leakage in the other direction.

Also worth adding coverage for whatever Trail Running / Track Running / Treadmill behavior is settled on per the comment at `coros_api.py:657`.

**Review by cygnusb — CHANGES_REQUESTED** [review]
Thanks — the core mapping and metadata block look solid, and the cycling regression test is reassuring. A few issues need fixing before merge.

## Summary

1. **Bug (correctness):** position-based `exerciseType` classification misclassifies sub-steps inside repeat groups as warmup/cooldown. Common interval shapes will render wrong on the watch. See inline on `coros_api.py:758`.
2. **Namespace coverage:** only activity-side `100` is mapped. `102` (Trail Running), `103` (Track Running), and any indoor-running / treadmill IDs silently fall through to a Running-but-no-metadata payload — the exact failure mode this PR fixes. Either map them too or reject them with a clear error. Inline on `coros_api.py:657`.
3. **Leaky abstraction:** `sport_type=1` (the wire ID) silently produces a run payload *without* the metadata block, because `is_running` keys off `100` only. Inline on `coros_api.py:658`.
4. **Test gap:** no test exercises a running workout containing a repeat group (the common interval shape) — closely related to #1. Inline on the tests file.
5. **Discoverability gap (server.py — outside this diff, please add):** the user-facing MCP tool docstrings + the `mcp_overview` entries still describe cycling-only. LLM callers won't discover that `sport_type=100` is now supported. Specifically:
   - `server.py:117` — overview entry for `save_workout_template` still says "cycling/intervals".
   - `server.py:121` — overview entry for `schedule_workout` still says "cycling/intervals".
   - `server.py:513` — `save_workout_template` docstring header still says "REUSABLE cycling/intervals workout TEMPLATE".
   - `server.py:561–562` — `sport_type` param docs only list 2 / 200 / 201; add `100 = Running` (and any Trail/Track/Treadmill IDs that end up supported per #2). Briefly note the activity↔workout namespace mapping so callers don't try wire `1`.
   - `server.py:737` — `schedule_workout` docstring header still says "ONE-OFF cycling/intervals workout".
   - `server.py:759–760` — `sport_type` param docs only mention Indoor Cycling; same additions as for `save_workout_template`.

Details inline.

File: coros_mcp/coros_api.py (hunk 1 above)
Comment (LeonLukaschek) [reply coros_mcp/coros_api.py]: Fixed in f7c0049. The warmup/cooldown markers are now restricted to top-level plain steps (`not isGroup and groupId == "0"`), so anything inside a repeat group keeps `exerciseType=2`. Both shapes from your comment are covered by new tests (`test_running_repeat_group_substeps_are_main`, `test_running_repeat_group_only_all_main`).

I also verified this on my watch (EU region): scheduled a 10min warmup, 3x(3min hard / 2min easy), 10min cooldown with HR targets through the patched builder. Opens fine in the app and renders as Running with the repeat group and HR zones intact.

File: coros_api.py (hunk 2 above)
Comment (LeonLukaschek) [reply coros_api.py]: Done, used exactly this predicate in f7c0049.

File: coros_api.py (hunk 3 above)
Comment (LeonLukaschek) [reply coros_api.py]: Fixed in f7c0049: added `_RUNNING_ACTIVITY_SPORT_TYPES = frozenset({100, 102, 103})`. Trail and Track now map onto wire `sportType=1` and get the full metadata block, with a parametrized test for both.

Regarding treadmill/indoor running: the codebase has no indoor-run activity ID anywhere (`SPORT_NAMES` doesn't list one) and I didn't want to map a guess I can't verify against the API. If you know the ID it's a one-line addition to the frozenset, either here or in a follow-up.

File: coros_mcp/coros_api.py (hunk 4 above)
Comment (LeonLukaschek) [reply coros_mcp/coros_api.py:803]: Fixed in f7c0049, went with the reject option. `sport_type=1` now raises `ValueError` with the wording you suggested (covered by `test_wire_sport_type_id_is_rejected`). The MCP tool layer wraps builder calls in try/except, so callers get a clean error dict pointing them at 100 instead of a crash.

File: coros_mcp/coros_api.py (hunk 5 above)
Comment (LeonLukaschek) [reply coros_mcp/coros_api.py:632]: Added in f7c0049, expanded to also mention 102/103 and why the old `100: "Running"` entry never round-tripped.

File: coros_mcp/coros_api.py (hunk 6 above)
Comment (LeonLukaschek) [reply coros_mcp/coros_api.py:778]: Done in f7c0049. The docstring now lists the supported IDs per namespace (2/200/201 for cycling, 100/102/103 for running), the 100 -> 1 wire mapping, and the rejection of wire ID 1. The server.py docstrings and mcp_overview entries from your summary are updated to match.

File: tests/test_workout_payloads.py (hunk 7 above)
Comment (LeonLukaschek) [reply tests/test_workout_payloads.py:430]: Added in f7c0049, all three:

1. `test_running_repeat_group_substeps_are_main`: [warm, repeat(3x[hard, easy]), cool] asserts exactly one `exerciseType=1`, one `3`, all sub-steps `2`, and that the sub-steps carry the per-step run metadata.
2. `test_running_repeat_group_only_all_main`: repeat-only shape, no 1/3 anywhere.
3. `test_cycling_payload_top_level_keys_are_exact`: cycling top-level key set is exactly `{name, sportType, estimatedTime, access, exercises}`.

Also added `test_trail_and_track_map_to_running` (parametrized over 102/103) and `test_wire_sport_type_id_is_rejected` for the namespace decisions.

**Comment by LeonLukaschek** [comment]
Thanks for the review. All five points are addressed in f7c0049, and the branch is rebased onto the v0.3 main.

1. Repeat-group misclassification: warmup/cooldown markers now only apply to top-level plain steps (`not isGroup and groupId == "0"`). Repeat sub-steps always get `exerciseType=2`.
2. Namespace coverage: Trail (102) and Track (103) map to wire `sportType=1` with the metadata block via `_RUNNING_ACTIVITY_SPORT_TYPES`. There is no indoor-run/treadmill ID anywhere in the codebase, so I left that out rather than guessing. Details in the inline thread.
3. Wire-ID leak: `sport_type=1` is rejected with a `ValueError` pointing at 100.
4. Tests: repeat-group classification (both shapes), trail/track mapping, wire-ID rejection, and a regression pinning the cycling payload's exact top-level key set.
5. Docs: the `mcp_overview` entries, both tool docstring headers, and both `sport_type` param blocks in server.py now cover running and the namespace mapping. `WORKOUT_SPORT_NAMES` and the builder docstring carry the namespace note.

On-device check (EU region, app + watch): scheduled a 10min warmup, 3x(3min hard / 2min easy), 10min cooldown with HR targets through the patched builder. Opens without crashing and renders as Running with the repeat group and HR zones intact. I also pushed a strength workout through the same build to make sure that path is untouched (bodyweight and kg weight encodings render correctly).

87/87 tests, ruff and `mypy . --exclude tests/` clean.

**Review by cygnusb — COMMENTED** [review]
## Code Review Findings

### Bug / Ambiguity: `exerciseNum` and `totalSets` include group containers

```python
"exerciseNum": len(exercises),
"totalSets": len(exercises),
```

`exercises` contains both real steps **and** the structural group container objects (entries with `isGroup=True`). A workout like `[warmup, repeat(3×2 sub-steps), cooldown]` produces 5 entries in `exercises` but only 4 of them are actual exercise steps (the group container is structural glue). If the Coros API counts only real steps in these fields, this is off-by-one for any running workout that contains a repeat group.

**Suggestion:** use the filtered count, or at minimum verify against a real API round-trip:

```python
real_step_count = sum(1 for e in exercises if not e.get("isGroup"))
payload.update({
    "exerciseNum": real_step_count,
    "totalSets": real_step_count,
    ...
})
```

Also: no test currently asserts the values of `exerciseNum`/`totalSets` for interval workouts — adding one would make the expected behaviour explicit.

---

### Missing entries in `WORKOUT_SPORT_NAMES`

```python
WORKOUT_SPORT_NAMES: dict[int, str] = {
    1: "Running",
    2: "Indoor Cycling",
    ...
    # 102, 103 are absent
}
```

Trail Running (`102`) and Track Running (`103`) are accepted by `_build_workout_program_payload` but are not in `WORKOUT_SPORT_NAMES`. Any code that calls `WORKOUT_SPORT_NAMES.get(sport_type)` with an activity-namespace ID before the namespace mapping occurs will get `None` for those types. `_parse_workout` uses this dict — if it ever receives a `sportType` from a running workout it will silently return `None` for the sport name.

**Suggestion:** either add the missing entries —

```python
WORKOUT_SPORT_NAMES: dict[int, str] = {
    1: "Running",
    2: "Indoor Cycling",
    4: "Strength",
    100: "Running",
    102: "Trail Running",
    103: "Track Running",
    200: "Road Bike",
    201: "Indoor Cycling (alt)",
}
```

— or make sure the dict is only ever consulted post-mapping (wire IDs only) and document that constraint explicitly.

---

### Minor: sub-step `exerciseType` set by coincidence, not by intent

In the running block, sub-steps inside repeat groups correctly end up with `exerciseType=2` (main), but only because the cycling-path construction code already wrote `"exerciseType": 2` for all non-group steps. The running metadata pass then skips group containers and does not touch this field on sub-steps.

This is correct today, but it's an implicit dependency between the cycling construction path and the running post-processing. A future refactor of the cycling path could silently break running interval workouts.

**Suggestion:** make the intent explicit with a comment, or add an explicit assignment in the per-step running loop:

```python
for ex in exercises:
    if ex.get("isGroup"):
        continue
    # exerciseType already set by position pass (top-level) or defaults to 2 (sub-steps)
    ex.setdefault("exerciseType", 2)   # sub-steps: always main
    ...
```

**Comment by LeonLukaschek** [comment]
Thanks for the second pass. All three are addressed in ed9c42e (pushed to the branch).

1. exerciseNum / totalSets counting the group container. Fixed: both now count non-group rows only (`real_step_count = sum(1 for e in exercises if not e.get("isGroup"))`). Flat workouts have no container rows so their counts are unchanged; the inflation only ever hit workouts with a repeat group. Added `test_running_repeat_group_counts_real_steps_only`, which pins 4 real steps for warmup / 3x(hard/easy) / cooldown even though `exercises` holds 5 rows.

2. WORKOUT_SPORT_NAMES missing 102/103. I went with the second option (document the constraint) rather than adding the entries. The dict is consumed only by `_parse_workout`, which reads workout-list API responses, and those carry workout-namespace (wire) sportType. The write path collapses every run flavor (100/102/103) to wire 1, so a run fetched back always reads as 1 and resolves to Running. 102/103 can't appear at that lookup. Adding activity-namespace IDs to a wire-keyed dict would be misleading rather than safer, so I documented the dict as wire-keyed (map activity IDs through _RUNNING_ACTIVITY_SPORT_TYPES first) and added a one-line note at the parse site. If you'd rather have the explicit entries anyway, say so and I'll add them.

3. Sub-step exerciseType set by coincidence. The running metadata pass now sets exerciseType=2 explicitly for repeat sub-steps (groupId != "0") instead of inheriting the value the cycling construction path writes, so a cycling-path refactor can't silently mislabel interval runs. Behavior is unchanged, so the existing repeat-group test still covers it.

88/88 tests, ruff and mypy . --exclude tests/ clean.

**Comment by cygnusb** [comment]
## Review findings

### Missing tests

**1. Two-step run (warmup + cooldown, no main block)**

With exactly 2 top-level plain steps, the current logic assigns `exerciseType=1` to the first and `exerciseType=3` to the second — producing a workout with no main step. It is unclear whether the COROS API accepts this. `test_running_hr_intensity_marks_hr_type` silently exercises this pattern but does not assert on `exerciseType`.

Suggested test:
```python
def test_running_two_step_exercise_types():
    """Two plain steps: first=warmup(1), second=cooldown(3), no main block."""
    payload = _build_workout_program_payload(
        name="r",
        steps=[
            {"name": "W", "duration_minutes": 5,  "intensity_low": 100, "intensity_high": 130},
            {"name": "C", "duration_minutes": 5,  "intensity_low": 100, "intensity_high": 130},
        ],
        sport_type=100,
        intensity_type=2,
    )
    assert [ex["exerciseType"] for ex in payload["exercises"]] == [1, 3]
```

If a 2-step workout is invalid on the wire, add a `ValueError` guard in `_build_workout_program_payload` for `is_running and len(top_level_plain) == 2`.

**2. Non-HR intensity types for running (pace, speed, cadence)**

There is no test confirming that `hrType=0` is emitted when `intensity_type != 2`. A parametrize would cover this cheaply:

```python
@pytest.mark.parametrize("intensity_type", [3, 4, 5, 7])  # pace, speed, none, cadence
def test_running_non_hr_intensity_emits_hr_type_zero(intensity_type):
    payload = _build_workout_program_payload(
        name="r",
        steps=[{"name": "Run", "duration_minutes": 30, "intensity_low": 0, "intensity_high": 0}],
        sport_type=100,
        intensity_type=intensity_type,
    )
    assert all(ex["hrType"] == 0 for ex in payload["exercises"])
    assert payload["referExercise"]["hrType"] == 0
```

**3. Road Bike (200) still unaffected**

`sport_type=2` (Indoor Cycling) is confirmed unchanged, but `sport_type=200` (Road Bike) takes a slightly different code path through the sport-type check. One test to confirm it also bypasses the running metadata block:

```python
def test_road_bike_not_treated_as_running():
    payload = _build_workout_program_payload(
        name="c",
        steps=[{"name": "ride", "duration_minutes": 60, "intensity_low": 200, "intensity_high": 250}],
        sport_type=200,
    )
    assert payload["sportType"] == 200
    assert "subType" not in payload
    assert "referExercise" not in payload
```

---

### Issues

**Unknown sport type passes through silently**

Any `sport_type` not in `_RUNNING_ACTIVITY_SPORT_TYPES` and not `1` is used as-is as the wire ID. Passing e.g. `sport_type=50` produces a cycling-shaped payload with `sportType=50` — likely an invalid wire value that would fail silently on the COROS side. Consider:

```python
_KNOWN_SPORT_TYPES = _RUNNING_ACTIVITY_SPORT_TYPES | {2, 200, 201}

if sport_type not in _KNOWN_SPORT_TYPES and sport_type != 1:
    raise ValueError(f"Unknown sport_type={sport_type}. ...")
```

Or at minimum document in the docstring that unknown types pass through as cycling-shaped payloads.

**Review by cygnusb — COMMENTED** [review]
Left the missing-tests and unknown-sport-type issues as PR comments. Otherwise solid — logic is correct and manually verified.

**Review by cygnusb — APPROVED** [review]
Re-reviewed after the latest commits — all findings from my earlier review are addressed:

1. ✅ warmup/cooldown markers restricted to top-level plain steps (`not isGroup and groupId == "0"`); repeat sub-steps stay `exerciseType=2`.
2. ✅ Namespace coverage via `_RUNNING_ACTIVITY_SPORT_TYPES = {100, 102, 103}`, all mapped to wire `sportType=1`.
3. ✅ wire ID `1` is rejected with a clear `ValueError`.
4. ✅ `exerciseNum` / `totalSets` count real steps only (group containers excluded).
5. ✅ sub-step `exerciseType=2` set explicitly, no longer an implicit dependency.
6. ✅ `server.py` docstrings + `get_help` overview updated for running.
7. ✅ Comprehensive tests, including repeat-group shapes and the cycling key-set regression.

Verified locally on the branch: `pytest` 88 passed, `ruff check .` clean, `mypy coros_mcp` clean. LGTM 🚀

**Comment by cygnusb** [comment]
Thanks for the reviews. The branch is rebased onto current `main` (v0.3, including the merged training-plan PRs) and the open points from your last round are addressed:

**Edge cases / validation**
- Unknown `sport_type` values are now rejected (`_KNOWN_SPORT_TYPES` = Running 100/102/103 + Cycling 2/200/201) instead of silently passing through as an invalid wire ID.
- New tests: two-step run (`[1,3]`), non-HR intensities emit `hrType=0` (parametrized over pace/speed/none/cadence), Road Bike (200) stays on the cycling path, and unknown `sport_type` is rejected.

**Additionally found & fixed in the re-review**
- *Correctness:* the warmup/cooldown markers were gated on a count of top-level plain steps, so `[warmup, intervals]` labeled the warmup as `exerciseType=2` (instead of `1`) — likewise `[intervals, cooldown]` labeled the cooldown as `2`. The markers now key off the position of the first/last top-level *item* (plain or group) and are only set when that item is a plain step.
- *UX:* `intensity_type` now resolves per sport when omitted — runs → 2 (HR), cycling → 6 (power). An explicit value still wins; cycling behavior is unchanged.

108 tests pass, `ruff` and `mypy . --exclude tests/` clean.
